services/server/src/server/controllers/repository/repository.handlers.ts

```diff
@@ -128,9 +136,71 @@ export async function checkAllByChainAndAddressEndpoint(
             });
           }
 
-          map
-            .get(address)
-            .chainIds.push({ chainId, status: found[0].runtimeMatch });
+          // Proxy detection and resolution
+          type Implementation = { address: string; name?: string };
+          let proxyStatus: {
+            isProxy?: boolean;
+            proxyType?: ProxyType | null;
+            implementations?: Implementation[];
+            proxyResolutionError?: string;
+          } = {};
+          if (
+            req.query.resolveProxies === "true" &&
+            found[0].onchainRuntimeBytecode
+          ) {
+            try {
+              const sourcifyChain = chainRepository.supportedChainMap[chainId];
+              const proxyDetectionResult = await detectAndResolveProxy(
+                found[0].onchainRuntimeBytecode,
+                address,
+                sourcifyChain,
+              );
+
+              const implementations = await Promise.all(
+                proxyDetectionResult.implementations.map(
+                  async (implementationAddress) => {
+                    implementationAddress = getAddress(implementationAddress);
+                    const implementation: Implementation = {
+                      address: implementationAddress,
+                    };
+
+                    const implementationFound: Match[] =
+                      await services.storage.performServiceOperation(
+                        "checkAllByChainAndAddress",
+                        [implementationAddress, chainId],
+                      );
+                    if (
+                      implementationFound.length > 0 &&
+                      implementationFound[0].contractName
+                    ) {
+                      implementation.name = implementationFound[0].contractName;
+                    }
+                    return implementation;
+                  },
+                ),
+              );
+
+              proxyStatus = { ...proxyDetectionResult, implementations };
+              logger.debug("Ran proxy detection and resolution", {
+                chainId,
+                address,
+                proxyStatus,
+              });
+            } catch (error) {
+              proxyStatus.proxyResolutionError = (error as Error)?.message;
+              logger.error("Error detecting and resolving proxy", {
+                chainId,
+                address,
+                error,
+              });
+            }
+          }
+
+          map.get(address).chainIds.push({
+            chainId,
+            status: found[0].runtimeMatch,
+            ...proxyStatus,
+          });
         }
       } catch (error) {
         // ignore
```

Comment:
Not introduced by this PR but not a good idea to ignore silently. We should log here